ArrayIndexOutOfBoundsException when serializing generic lambda written in Kotlin

**Describe the bug**
When serializing an object in Kotlin it fails with an ArrayIndexOutOfBoundsException
```
Exception in thread "main" com.esotericsoftware.kryo.KryoException: java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 1
Serialization trace:
definition (org.koin.core.definition.BeanDefinition)
	at com.esotericsoftware.kryo.serializers.ReflectField.write(ReflectField.java:101)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:108)
	at com.esotericsoftware.kryo.Kryo.writeClassAndObject(Kryo.java:711)
	at KryoBeanDefinitionKt.main(KryoBeanDefinition.kt:31)
	at KryoBeanDefinitionKt.main(KryoBeanDefinition.kt)
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 1
	at com.esotericsoftware.kryo.util.DefaultGenerics.pushTypeVariables(DefaultGenerics.java:116)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.pushTypeVariables(FieldSerializer.java:147)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:102)
	at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:642)
	at com.esotericsoftware.kryo.serializers.ReflectField.write(ReflectField.java:70)
	... 4 more

Process finished with exit code 1
```

**To Reproduce**
I've created a minimal example in https://github.com/holgerbrandl/kryo-koin. The respective class to reproduce the stacktrace from above is `KryoBeanDefinitionKt#main`


**Environment:**
 - OS: Windows 10
 - JDK Version: 11
 - Kryo Version: 5.2.0

**Analysis**
It seems to be a mismatch between the array sizes of `args` and `hierarchy` in  
https://github.com/EsotericSoftware/kryo/blob/76c5e5d967a8d4a7e438ffcc4dd0be825ca93ce3/src/com/esotericsoftware/kryo/util/DefaultGenerics.java#L116
![image](https://user-images.githubusercontent.com/200952/140645833-541769ad-9378-4aa5-b0e2-6f07b5e1b5b5.png)

I've tried, but did not manage to narrow down the problem any further.

--- a/src/com/esotericsoftware/kryo/util/DefaultGenerics.java
+++ b/src/com/esotericsoftware/kryo/util/DefaultGenerics.java
@@ -93,8 +93,9 @@ public final class DefaultGenerics implements Generics {
 
 	@Override
 	public int pushTypeVariables (GenericsHierarchy hierarchy, GenericType[] args) {
-		// Do not store type variables if hierarchy is empty or we do not have arguments for all root parameters.
-		if (hierarchy.total == 0 || hierarchy.rootTotal > args.length) return 0;
+		// Do not store type variables if hierarchy is empty, or we do not have arguments for all root parameters, or we have more
+		// arguments than the hierarchy has parameters.
+		if (hierarchy.total == 0 || hierarchy.rootTotal > args.length || args.length > hierarchy.counts.length) return 0;
 
 		int startSize = this.argumentsSize;
 

--- a/test-kotlin/com.esotericsoftware.kryo/SerializationTest.kt
+++ b/test-kotlin/com.esotericsoftware.kryo/SerializationTest.kt
@@ -4,16 +4,12 @@ import com.esotericsoftware.kryo.io.Input
 import com.esotericsoftware.kryo.io.Output
 import org.junit.jupiter.api.Assertions.assertEquals
 import org.junit.jupiter.api.Assertions.assertNotSame
-import org.junit.jupiter.api.Disabled
 import org.junit.jupiter.api.Test
 
-class LambdaTest {
-    class Example(private val p: (Long) -> String) {
-        constructor() : this({ it.toString() })
-    }
+class SerializationTest {
 
+    // https://github.com/EsotericSoftware/kryo/issues/864
     @Test
-    @Disabled("Expected to fail")
     fun testLambda() {
         val kryo = Kryo().apply {
             isRegistrationRequired = false
@@ -32,6 +28,10 @@ class LambdaTest {
 
         assertEquals(example::class.java, deserialized::class.java)
         assertNotSame(example, deserialized)
+    }
+
+    class Example(private val p: (Long) -> String) {
+        constructor() : this({ it.toString() })
     }
 }
 

